### ISSUE TITLE
emit diagnostic on access of `NotRequired` typeddict key

### ISSUE THREAD

**Issue by GniLudio** [astral-sh/ty#2810]
### Summary

Even though the [feature overview](https://github.com/astral-sh/ty/issues/1889) and the [support TypedDict](https://github.com/astral-sh/ty/issues/154) issues show `NotRequired` as supported, it seems to be not working correctly.

In the following example `MyDict.foo` must not necessarily exist, but `ty` always assumes it's existence.

## Example
```python
from typing import TypedDict, NotRequired

class MyDict(TypedDict):
    foo: NotRequired[str]

my_dict: MyDict = {}

# assignment to a string variable doesn't give a warning
my_string: str = my_dict["foo"]

# assignment to a string parameter doesn't give a warning
def print_string(s: str) -> None:
    print(s)
print_string(my_dict["foo"])
```
[Playground](https://play.ty.dev/c2aa0c38-b082-43e6-bd8c-8ab423478df0)

### Version

ty 0.0.17 (8cec85718 2026-02-13)

**Comment by carljm** [comment]
Yes, we should add this diagnostic - thanks for the report. (Only pyright seems to emit a diagnostic in these cases; mypy and pyrefly do not. So there doesn't seem to be a consistent interpretation of type checkers here. But clearly soundness requires it.)

**Comment by AlexWaygood** [comment]
Mypy's had a PR open for this since 2023, but with little movement, I think because of the number of ecosystem hits: https://github.com/python/mypy/pull/14717.

While I agree that this check would be useful, we could also consider making the check disabled by default. I don't think soundness "clearly" requires this — for a `dict[int, int]` object `x`, we would obviously not emit a diagnostic on `x[0]`, even though the type of `x` does not guarantee that `x` actually has a `0` key in it.

**Comment by GniLudio** [comment]
I don't know whether it's related to this issue or a separate one, but the type of `get` also isn't correct.

## Example
```python
from typing import TypedDict, NotRequired

class MyDict(TypedDict):
    name: str
    foo: NotRequired[int]


my_dict: MyDict = {
    "name": "John", 
    "foo": 5,
}

# detected correctly as 'str'
name = my_dict.get("name") 

# detected as 'Unknown | default'
foo_1 = my_dict.get("hello") 
foo_2 = my_dict.get("hello", 5)
```
[Playground](https://play.ty.dev/fd85bc19-854a-4ac7-85d7-e58601b861fa)

**Comment by GniLudio** [comment]
> Mypy's had a PR open for this since 2023, but with little movement, I think because of the number of ecosystem hits: [python/mypy#14717](https://github.com/python/mypy/pull/14717).
> 
> While I agree that this check would be useful, we could also consider making the check disabled by default. I don't think soundness "clearly" requires this — for a `dict[int, int]` object `x`, we would obviously not emit a diagnostic on `x[0]`, even though the type of `x` does not guarantee that `x` actually has a `0` key in it.

I agree for `dict`, while thinking that for things like `TypedDict` (dataclasses, pydantic, etc.) it should be checked by default. 
When explicitly defining the types, I expect the typer checker to validate them (to some degree).

**Comment by carljm** [comment]
> While I agree that this check would be useful, we could also consider making the check disabled by default. I don't think soundness "clearly" requires this — for a `dict[int, int]` object `x`, we would obviously not emit a diagnostic on `x[0]`, even though the type of `x` does not guarantee that `x` actually has a `0` key in it.

Ok, "clearly" was maybe overstating it -- but I agree with @GniLudio that it's different for TypedDict, when you've explicitly said in the type that a given key might be absent, and we will error for accessing keys that you didn't mention at all.

I don't feel super strongly about whether this rule would be enabled by default or not, as long as it's available for those who want it. It's the sort of thing that might belong in a strict mode.

**Comment by carljm** [comment]
@GniLudio The question about `.get` return types definitely belongs in a separate issue from this one. (I don't think ty's inference here is wrong, but it is more forgiving -- the other, stricter, option would be `object`. Both options are "correct", one is just stricter and the other more gradual. Pyrefly and mypy use `object`, pyright uses `Any | default`. If you'd like to open an issue to advocate for ty to use `object`, feel free!)

**Comment by DetachHead** [comment]
> Yes, we should add this diagnostic - thanks for the report. (Only pyright seems to emit a diagnostic in these cases; mypy and pyrefly do not. So there doesn't seem to be a consistent interpretation of type checkers here. But clearly soundness requires it.)

[pyrefly now does](https://pyrefly.org/sandbox/?project=N4IgZglgNgpgziAXKOBDAdgEwEYHsAeAdAA4CeS4ATrgLYAEALqcROgOZ0Q3G6UN0AVZjEwARCAGMGAGjoA5XAwBKMAI4BXCJREAdSjvR6DEqKjhw6AWVLipACiHERthgEpER9HW90wuXIjyiioaWiIA2nAMlAC6np40pAD6mJIMgdYudAC8dMAAvvH66ADEdGZwEGzoNDDo-Ay45XRRlKwcAG6obajYsHSYuPDoAOT8bBAdMM0A7t3o7QnJre2BrTl0iSlp4Trg-ntxxZ5lFVU1dQ1NqC3R7XTE3ai1DDCUA0Nwo%2BOT0zdzlAW7E8mBgYAebXqSRW7DscDW0VcdAAtAA%2BILoGAeYo%2BCGsBhw1yeYiQhjQu6wrapKS7fa4Q6uEDSEBkbRgKCkQiNGhQChlAAKpDZHJaGBwBDoElw6EgbHUlFQDAg0sIBjKAGUYNMABYMBjEeEAekNrLBHMIvDYhrqhsGEjghqlMqq8sVyvQht8vHKXWgvX6TtlrqV0rouGIIfQcFV6DIDG10uRU0olVDuT2AGZCABGABMewM4Te1BTcXQ6gW3F4r0wyNS2ikvw2I3OvBgIwM6EUyO0oW0tYA1jBSMjUBIJPALLkRgDRgYuMQoJIIAxR%2BhSM3i7wO5j8Ivl6uMBvp1vKDuF0uJCuOXWwawRKO9W1sOpXpvKCWRiB8syx0qpgAYtAMAUGgWB4EQZDfkAA)
